shell/platform/windows/windows_lifecycle_manager.h

```diff
@@ -0,0 +1,57 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_WINDOWS_WINDOWS_LIFECYCLE_MANAGER_H_
+#define FLUTTER_SHELL_PLATFORM_WINDOWS_WINDOWS_LIFECYCLE_MANAGER_H_
+
+#include <Windows.h>
+
+#include <cstdint>
+#include <map>
+#include <optional>
+
+namespace flutter {
+
+class FlutterWindowsEngine;
+
+/// A manager for lifecycle events of the top-level window.
+///
+/// Currently handles the following events:
+/// WM_CLOSE
+class WindowsLifecycleManager {
+ public:
+  WindowsLifecycleManager(FlutterWindowsEngine* engine);
+  virtual ~WindowsLifecycleManager();
+
+  // Called when the engine is notified it should quit, e.g. by an application
+  // call to `exitApplication`. When window is std::nullopt, this quits the
+  // application. Otherwise, it holds the HWND of the window that initiated the
+  // request, and exit_code is unused.
+  virtual void Quit(std::optional<HWND> window,
+                    std::optional<WPARAM> wparam,
+                    std::optional<LPARAM> lparam,
+                    UINT exit_code);
```

Comment:
What do you think of having two separate APIs here?

1. Close window
2. Quit app

Apologies for not suggesting this earlier, didn't think of this until now :)

Reply:
We currently intend for the two to both be controlled by the `didRequestAppExit` framework-side callback, which is why they are handled by the same engine-side method as well.